### DISCUSSION TITLE
CombineNavigation package for UIKit

### DISCUSSION THREAD

**Discussion by maximkrouk** [pointfreeco/swift-composable-architecture#2662] [Show and tell]
Finally finished implementing core logic for [https://github.com/capturecontext/combine-cocoa-navigation](https://github.com/capturecontext/combine-cocoa-navigation/tree/navigation-stacks)

I still need to finish the example, cleanup/improve readme and docs, but existing tests are passing and simple example seems to work perfectly.

Hope to release `1.0.0` by the end of this year, so if someone finds time to play with it before the release, I would love to hear your feedback ❤️

Lets make TCA with UIKit great at the first time 😁

**Comment by maximkrouk** [comment]
Created `observation-beta` branch in [`composable-architecture-extensions`](https://github.com/capturecontext/composable-architecture-extensions/tree/observation-beta) and updated [`https://github.com/capturecontext/combine-cocoa-navigation`](https://github.com/capturecontext/combine-cocoa-navigation/tree/navigation-stacks) example using extensions package

The example is yet pretty basic, but now navigation API is tailored for TCA, not just Combine

**Reply by maximkrouk** [reply]
Some `observation-beta` feedback:

-----

I noticed that @ObservableState does not support typealiases like
```swift
@ObservableState
typealias State = IdentifiedArrayOf<OtherFeature.State>
```
That's kinda unfortunate, but don't see how it can be implemented since macro needs to add some props to the declaration 💁‍♂️

-----

`@Reducer` adds `@CasePathable` macro to `Action` but doesn't add `@ObservableState` to `State`, not sure if it was intended by design, but confused me a little, I may create a PR if it's a bug 🫠

-----

@stephencelis @mbrandonw observation-beta discussion is closed and the feedback feels too minor to create a separate discussion, so I decided to provide it here 🌚

**Reply by mbrandonw** [reply]
Hi @maximkrouk, definitely feel free to open a new discussion. It's easier for us to track than adding it to an unrelated thread.

> I noticed that @ObservableState does not support typealiases like

That is just a limitation of macros, and the same is true of `@Observable`. Also, `IdentifiedArray` is not observable.

> @Reducer adds @CasePathable macro to Action but doesn't add @ObservableState to State, not sure if it was intended by design, but confused me a little, I may create a PR if it's a bug 🫠

It's a purposeful omission for now. Suddenly adding `@ObservableState` to _everyone's_ state has huge ramifications. It will lead to over observation of state when mixed with `WithViewStore`, and it will lead to lots of runtime warnings telling people that they need to wrap their views in `WithPerceptionTracking` for iOS <17 support.

Once we can fully remove the concept of `ViewStore` from the library and force everyone to use `@ObservableState`, then we can have `@Reducer` automatically apply the macro.

**Reply by maximkrouk** [reply]
> Definitely feel free to open a new discussion. It's easier for us to track than adding it to an unrelated thread.

Ok, I'll be creating new discussions if I find smth else to share 👌

> That is just a limitation of macros, and the same is true of @Observable

Yeah kinda makes sense 🫠

> It's a purposeful omission for now. .....

Thank you for clarification 👍

**Reply by maximkrouk** [reply]
Just watched available `ObservableArchitecture` episodes, very cool step-by-step guide on "hows and whyes" of most recent changes in the repo ❤️

**Comment by maximkrouk** [comment]
Since there is a swift-navigation library with (imo) better APIs and first-class support in TCA, I'm closing the discussion :)
